src/jit/gentree.cpp

```diff
@@ -16563,6 +16563,114 @@ CORINFO_CLASS_HANDLE Compiler::gtGetClassHandle(GenTree* tree, bool* isExact, bo
+                        *isNonNull = helperResultNonNull;
+                        break;
+                    }
+
+                    case CORINFO_HELP_CHKCASTCLASS:
+                    case CORINFO_HELP_CHKCASTANY:
+                    case CORINFO_HELP_CHKCASTARRAY:
+                    case CORINFO_HELP_CHKCASTINTERFACE:
+                    case CORINFO_HELP_CHKCASTCLASS_SPECIAL:
+                        isCastHelper = true;
+                        __fallthrough;
+
+                    case CORINFO_HELP_ISINSTANCEOFINTERFACE:
+                    case CORINFO_HELP_ISINSTANCEOFARRAY:
+                    case CORINFO_HELP_ISINSTANCEOFCLASS:
+                    case CORINFO_HELP_ISINSTANCEOFANY:
+                    {
+                        // Fetch the class handle from the helper call arglist
+                        GenTreeArgList*      args    = call->gtCallArgs;
+                        GenTree*             typeArg = args->Current();
+                        CORINFO_CLASS_HANDLE castHnd = gtGetHelperArgClassHandle(typeArg);
+
+                        // We generally assume the type being cast to the best type
+                        // for the result, unless it is an interface type.
+                        //
+                        // Todo: when we have default interface methods then this
+                        // might not be the best assumption.
+                        if (castHnd != nullptr)
+                        {
+                            DWORD attrs = info.compCompHnd->getClassAttribs(castHnd);
+
+                            if ((attrs & CORINFO_FLG_INTERFACE) != 0)
+                            {
+                                castHnd = nullptr;
+                            }
+                        }
+
+                        // If we don't have a good estimate for the type we can use the
+                        // type from the value being cast instead.
+                        if (castHnd == nullptr)
+                        {
+                            GenTree* valueArg = args->Rest()->Current();
+                            castHnd           = gtGetClassHandle(valueArg, isExact, isNonNull);
+                        }
+
+                        // We don't know at jit time if the cast will succeed or fail, but if it
+                        // fails at runtime then an exception is thrown for cast helpers, or the
+                        // result is set null for instance helpers.
+                        //
+                        // So it safe to claim the result has the cast type, and that if we have a
+                        // cast the result is not null.
+                        //
+                        // Note we don't know for sure that it is exactly this type.
+                        if (castHnd != nullptr)
+                        {
+                            objClass   = castHnd;
+                            *isExact   = false;
+                            *isNonNull = isCastHelper;
```

Comment:
`CORINFO_HELP_CHKCAST*` methods guarantee a non-null result only if the result type is a struct.